evml/keras/models.py

```diff
@@ -1,49 +1,132 @@
+        """
         
-        for i in range(num_hidden_layers):
-            if num_hidden_layers == 1:
-                model.add(tf.keras.layers.Dense(hidden_size))
-                model.add(tf.keras.layers.LeakyReLU())
-            else:
-                model.add(tf.keras.layers.Dense(hidden_size))
-                model.add(tf.keras.layers.LeakyReLU())
-                model.add(tf.keras.layers.Dropout(dropout_rate))
-    else:
-        model.add(tf.keras.layers.Dense(input_size, activation=activation))
+        nn_input = Input(shape=(inputs,), name="input")
+        nn_model = nn_input
         
-        for i in range(num_hidden_layers):
-            if num_hidden_layers == 1:
-                model.add(tf.keras.layers.Dense(hidden_size, activation=activation))
-            else:
-                model.add(tf.keras.layers.Dense(hidden_size, activation=activation))
-                model.add(tf.keras.layers.Dropout(dropout_rate))
-      
-    model.add(tf.keras.layers.Dense(output_size, activation=out_activation))
-    
-    return model
+        if self.activation == 'leaky':
+            self.activation = LeakyReLU()
+        
+        if self.kernel_reg == 'l1':
+            self.kernel_reg = L1(self.l1_weight)
+        elif self.kernel_reg == 'l2':
+            self.kernel_reg = L2(self.l2_weight)
+        elif self.kernel_reg == 'l1_l2':
+            self.kernel_reg = L1L2(self.l1_weight, self.l2_weight)
+        else:
+            self.kernel_reg = None
+        
+        for h in range(self.hidden_layers):
+            nn_model = Dense(self.hidden_neurons, activation=self.activation,
+                             kernel_regularizer=L2(self.l2_weight), name=f"dense_{h:02d}")(nn_model)
+            if self.use_dropout:
+                nn_model = Dropout(self.dropout_alpha, name=f"dropout_h_{h:02d}")(nn_model)
+            if self.use_noise:
+                nn_model = GaussianNoise(self.noise_sd, name=f"ganoise_h_{h:02d}")(nn_model)
+        nn_model = DenseNormalGamma(outputs, name='DenseNormalGamma')(nn_model)
+        self.model = Model(nn_input, nn_model)
+        if self.optimizer == "adam":
+            self.optimizer_obj = Adam(lr=self.lr, beta_1=self.adam_beta_1, beta_2=self.adam_beta_2)
+        elif self.optimizer == "sgd":
+            self.optimizer_obj = SGD(lr=self.lr, momentum=self.sgd_momentum)
+        self.model.compile(optimizer=self.optimizer_obj, loss=self.loss,
+                           loss_weights=self.loss_weights, run_eagerly=False)
 
+    def fit(self, x, y):
+        inputs = x.shape[1]
+        if len(y.shape) == 1:
+            outputs = 1
+        else:
+            outputs = y.shape[1]
+        self.build_neural_network(inputs, outputs)
+        self.model.fit(x, y, batch_size=self.batch_size, epochs=self.epochs, verbose=self.verbose, shuffle=True)
+        self.training_var = np.var(y)
+        tf.keras.models.save_model(self.model, self.save_path+'/test_model/model.h5', save_format='h5')
```

Comment:
I suggest moving the save_model command out of the fit function and into a separate save_model method. This could cause hard-to-trace bugs for multiple reasons.
1. If `self.save_path` does not have a `/` at the end, you will get a path error.
2. If the test_model directory does not exist, you will get an OS Error. 
3. model.h5 is a very generic name. We should have some ability to customize model names or use something that produces a unique model id based on when the model was trained.

When constructing a path for saving anything, use os.path.join to link multiple sub-paths together into a larger path. It automatically handles the presence or absence of slashes.

Reply:
edited.